pymeasure/instruments/channel.py

```diff
@@ -0,0 +1,103 @@
+#
+# This file is part of the PyMeasure package.
+#
+# Copyright (c) 2013-2022 PyMeasure Developers
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+#
+
+import logging
+
+from .common_base import CommonBase
+
+log = logging.getLogger(__name__)
+log.addHandler(logging.NullHandler())
+
+
+class Channel(CommonBase):
+    """The base class for channel definitions.
+
+    This class supports dynamic properties like :class:`Instrument`,
+    but requires an :class:`Instrument` instance as a parent for communication.
+    The default implementation of :meth:`write` uses :code:`str.format` to
+    exchange :code:`'{ch}'` for the channel name.
+
+    :param parent: The instrument (an instance of :class:`~pymeasure.instruments.Instrument`)
+        to which the channel belongs.
+    :param id: Identifier of the channel, as it is used for the communication.
+    """
+
+    def __init__(self, parent, id):
```

Comment:
Huh, good thinking on coming up with that problem/use case! So, in your VNA, you could have commands like `{port}:{ch}:VOLT:{trace} %s`, is something like that what you are thinking of? If yes, customizable names sound useful.

Equating it to the prefix sounds nice, as it would avoid duplication. However, we are putting those placeholder strings in the command definitions (`bla{ch}34`), and the prefixes are in principle flexible/dynamic, so that would quickly get out of sync. For the same reason, a kwarg does not seem useful.

What about having the placeholder be an attribute, instead, with default `ch`?
 
Then, in your `MyTrace(Channel)`'s init, you'd say `self.ch_placeholder = 'trace'`, and the above command would be `:VOLT:{trace} %s` (the trace thing does not need to know about the channels), no need to override the write methods.
In your `MyChannel(Channel)` that contains all those traces, you'd leave the default placeholder in place, and override the 2 write methods to prefix the `MyChannel.ch` to all commands.

Reply:
Right now, you can choose the placeholder if you redefine the `write` method.
I like the idea to bind the placeholder to the prefix, but I see the point, that we have to sync all the command strings to the placeholder.

By the way, the AWG 401 changed in this PR (and I think the tests) has already nested channels (albeit just 2 deep).

Reply:
> is something like that what you are thinking of?

Indeed, see the command I added to another reply.

> What about having the placeholder be an attribute, instead, with default `ch`?

Good point about the flexibility. Defining it in the init is probably a better solution.

> (Not sure how the ports fit in here, so did not add those)

Ports are mostly parallel to the traces (e.g. `SOUR{ch}:POW:PORT{pt}`) although I think I also saw some commands which nests it within a trace, but I can't seem to find that one right now)

Reply:
> Right now, you can choose the placeholder if you redefine the `write` method.

However, if the placeholder is an attribute (and you don't need to do anything fancy), you _just_ redefine the attribute, and don't have to override two methods, as the write method could just use the instance attribute as the key in `format`:
```
self.parent.write(command.format(**{self.placeholder: self.id}), **kwargs)
``` 
(did not test that though :sweat_smile: )

Reply:
We have to introduce a single formatting method in case somebody wants to add a channel prefix instead of replacing a placeholder.
Now, you have only one place to change the name, however, I see your point of defining the placeholder at class level.

Reply:
> We have to introduce a single formatting method in case somebody wants to add a channel prefix instead of replacing a placeholder.

Yeah, I did not want to drag the (useful, but in principle unrelated) proposed `format_command` into this discussion, so kept it with the `write` example.

Reply:
Implemented.